src/storage/src/hummock/iterator/concat_inner.rs

```diff
@@ -100,21 +112,65 @@ impl<TI: SstableIteratorType> ConcatIteratorInner<TI> {
 impl<TI: SstableIteratorType> HummockIterator for ConcatIteratorInner<TI> {
     type Direction = TI::Direction;
 
+    type AwaitNextFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
     type NextFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
     type RewindFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
     type SeekFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
 
     fn next(&mut self) -> Self::NextFuture<'_> {
         async move {
-            let sstable_iter = self.sstable_iter.as_mut().expect("no table iter");
-            sstable_iter.next().await?;
+            match self.poll_next() {
+                Poll::Ready(result) => result,
+                Poll::Pending => self.await_next().await,
+            }
+        }
+    }
 
-            if sstable_iter.is_valid() {
-                Ok(())
-            } else {
-                // seek to next table
-                self.seek_idx(self.cur_idx + 1, None).await
+    fn poll_next(&mut self) -> Poll<HummockResult<()>> {
+        debug_assert_matches!(self.pending_stage, ConcatIteratorPendingStage::None);
+        let sstable_iter = self.sstable_iter.as_mut().expect("no table iter");
+        match sstable_iter.poll_next() {
+            Poll::Ready(result) => {
+                if result.is_err() {
+                    Poll::Ready(result)
+                } else if sstable_iter.is_valid() {
+                    Poll::Ready(Ok(()))
+                } else {
+                    // will seek to next table in await
+                    self.pending_stage = ConcatIteratorPendingStage::AwaitSeekIdx;
+                    Poll::Pending
+                }
             }
+            Poll::Pending => {
+                self.pending_stage = ConcatIteratorPendingStage::AwaitNext;
+                Poll::Pending
+            }
+        }
+    }
+
+    fn await_next(&mut self) -> Self::AwaitNextFuture<'_> {
+        async move {
+            let ret = match self.pending_stage {
+                ConcatIteratorPendingStage::None => Ok(()),
+                ConcatIteratorPendingStage::AwaitNext => {
+                    let sstable_iter = self.sstable_iter.as_mut().expect("should have table iter");
+                    match sstable_iter.await_next().await {
+                        Ok(()) => {
+                            if sstable_iter.is_valid() {
+                                Ok(())
+                            } else {
+                                self.seek_idx(self.cur_idx + 1, None).await
+                            }
+                        }
+                        Err(e) => Err(e),
+                    }
+                }
+                ConcatIteratorPendingStage::AwaitSeekIdx => {
+                    self.seek_idx(self.cur_idx + 1, None).await
```

Comment:
You should either implement the whole function as async state machine or using async function. I rarely see people mix these two. Not sure whether there will be incorrectness from such implementation.